File: stacks_and_queues/questions/question_3.4/problem_3.4.rb
class Node
  attr_accessor :value, :next

  def initialize value=nil
    @value = value
    @next = nil
  end
end

class Stack

  attr_accessor :top

  def initialize top=nil
    @top = Node.new top
  end

  def pop
    return nil unless top
    pop_result = top.value
    @top = top.next
    pop_result
  end

  def push item
    item = Node.new(item)
    item.next = top
    @top = item
  end

  def peek
    top.value
  end
end

class Disc
  attr_reader :size

  def initialize size
    @size = size
  end
end

class Tower
  attr_reader :tower_1, :tower_2, :tower_3

  def initialize num_discs=9
    @tower_1 = Stack.new
    @tower_2 = Stack.new
    @tower_3 = Stack.new
    @discs = []
    create_discs num_discs
    load_first_tower_with_discs
  end

  def solve
    move @discs.count, @tower_1, @tower_2, @tower_3
  end

  def move num_discs, from, pivot, destination
    # move num_discs -1 from from to pivot
    #  move last disc from from to destination
    #  move num_discs -1 from pivot to destination
    if num_discs == 1
      destination.push from.pop
    else
      move num_discs -1, from, destination, pivot
      move 1, from, pivot, destination
      move num_discs -1, pivot, from, destination
    end
  end


  private
  def create_discs num_discs
    num_discs.times {|size| @discs << Disc.new(size)}
  end

  def load_first_tower_with_discs
    @discs.reverse.each {|disc| @tower_1.push disc}
  end


end


tower = Tower.new
p tower.tower_1.peek
tower.solve
p tower.tower_1.peek
p tower.tower_3.peek

File: stacks_and_queues/questions/question_3.4/problem_3.4.js
var Node = function(value){
  timer = Date.now()
  while (Date.now() === timer){}
  this.value = value
  this.id = Date.now()
  this.next = null
};

var Stack = function(value){
  if (value) {this.top = new Node(value)}
};

Stack.prototype.pop = function(){
  if (!this.top) {return null}
  result = this.top
  this.top = this.top.next
  return result.value
}

Stack.prototype.push = function(item){
  nodeItem = new Node(item)
  nodeItem.next = this.top
  this.top = nodeItem
}

Stack.prototype.peek = function(){
  return (this.top ? this.top.value : null)
}

var TowerOfHanoi = function(numDiscs){
  this.numDiscs = (numDiscs = typeof numDiscs !== 'undefined' ? numDiscs : 7)
  this.tower1 = new Stack()
  this.tower2 = new Stack()
  this.tower3 = new Stack()

  for(var size = numDiscs; size > 0; size--){
    this.tower1.push(size)
  }
}

TowerOfHanoi.prototype.solve = function(){
  console.log(typeof(this.tower1))
  this.move(  this.numDiscs,
              this.tower1,
              this.tower2,
              this.tower3)
  // tests that tower 3 now contains all discs in correct order
  // console.log('contents of tower3: ')
  // while(this.tower3.peek()){console.log(this.tower3.pop())}
}

TowerOfHanoi.prototype.move = function(numDiscs, origin, pivot, destination){
  if(numDiscs == 1){
    destination.push(origin.pop())
    return
  }
  this.move(    numDiscs-1,
                origin,
                destination,
                pivot)
  this.move(    1,
                origin,
                pivot,
                destination)
  this.move(    numDiscs -1,
                pivot,
                origin,
                destination)
}


// ***************************************************************************
th = new TowerOfHanoi(9)
// while(th.tower1.peek()){console.log(th.tower1.pop())}
th.solve()
// th.tower3.push(th.tower1.pop())
